pkg/reconciler/kubequota/kubequota_controller.go

```diff
@@ -0,0 +1,352 @@
+		return fmt.Errorf("error starting quota controller for cluster %q: %w", clusterName, err)
+	}
+
+	return nil
+}
+
+func (c *Controller) startQuotaForClusterWorkspace(ctx context.Context, clusterName logicalcluster.Name) error {
+	resourceQuotaControllerClient := c.kubeClusterClient.Cluster(clusterName)
+	resourceQuotaControllerDiscoveryClient := resourceQuotaControllerClient.Discovery()
+
+	discoveryFunc := resourceQuotaControllerDiscoveryClient.ServerPreferredNamespacedResources
+
+	dynamicClient := c.dynamicClusterClient.Cluster(clusterName)
+
+	// TODO(ncdc): pass in a scoped factory when available, rather than creating a new factory per logical cluster.
+	// TODO(ncdc): could this be a partial metadata factory for everything except the few types that require full
+	// object details?
+	dynamicInformerFactory := dynamicinformer.NewFilteredDynamicSharedInformerFactoryWithOptions(
+		dynamicClient,
+		"",
+		nil,
+		cache.WithResyncPeriod(c.fullResyncPeriod),
+		cache.WithIndexers(cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc}),
+	)
+	dynamicInformerFactory.Start(ctx.Done())
+
+	conformingDynamicInformerFactory := &dynamicInformerFactoryAsNormalInformerFactory{
+		dynamicInformerFactory: dynamicInformerFactory,
+	}
+
+	listerFuncForResource := generic.ListerFuncForResourceFunc(conformingDynamicInformerFactory.ForResource)
+
+	quotaConfiguration := install.NewQuotaConfigurationForControllers(listerFuncForResource)
+
+	resourceQuotaControllerOptions := &resourcequota.ControllerOptions{
+		QuotaClient: resourceQuotaControllerClient.CoreV1(),
+		ResourceQuotaInformer: &SingleClusterResourceQuotaInformer{
+			clusterName:     clusterName,
+			delegate:        c.kubeInformers.Core().V1().ResourceQuotas(),
+			registerHandler: c.registerResourceQuotaEventHandlerLocked,
+		},
+		ResyncPeriod:    controller.StaticResyncPeriodFunc(c.quotaRecalculationPeriod),
+		InformerFactory: conformingDynamicInformerFactory,
+		ReplenishmentResyncPeriod: func() time.Duration {
+			return c.fullResyncPeriod
+		},
+		DiscoveryFunc:        discoveryFunc,
+		IgnoredResourcesFunc: quotaConfiguration.IgnoredResources,
+		InformersStarted:     c.informersStarted,
+		Registry:             generic.NewRegistry(quotaConfiguration.Evaluators()),
+	}
+	if resourceQuotaControllerClient.CoreV1().RESTClient().GetRateLimiter() != nil {
+		if err := ratelimiter.RegisterMetricAndTrackRateLimiterUsage(clusterName.String()+"-resource_quota_controller", resourceQuotaControllerClient.CoreV1().RESTClient().GetRateLimiter()); err != nil {
+			return err
+		}
+	}
+
+	resourceQuotaController, err := resourcequota.NewController(resourceQuotaControllerOptions)
+	if err != nil {
+		return err
+	}
+
+	go resourceQuotaController.Run(ctx, c.workersPerLogicalCluster)
+
+	// Periodically the quota controller to detect new resource types
+	go resourceQuotaController.Sync(discoveryFunc, 30*time.Second, ctx.Done())
```

Comment:
this is also by workspace? This does not scale :-/

Reply:
Yes, same comments as above & in Slack. Can come up with a scaling plan after we get this in, if that's ok.